src/snapshot/checksum.rs

```diff
@@ -0,0 +1,59 @@
+use std::{hash::Hash, marker::PhantomData};
+
+use bevy::prelude::*;
+
+use crate::{SaveWorld, SaveWorldSet};
+
+/// Flags an entity as containing a checksum for a type `T`
+#[derive(Component)]
+pub struct ChecksumFlag<T> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Default for ChecksumFlag<T> {
+    fn default() -> Self {
+        Self {
+            _phantom: default(),
+        }
+    }
+}
+
+/// Represents a checksum value for a specific type, flagged by [`ChecksumFlag`].
+#[derive(Component, Default, Hash)]
+pub struct ChecksumPart(pub u128);
+
+/// Represents a total checksum for a given frame.
+#[derive(Resource, Default, Clone, Copy)]
+pub struct Checksum(pub u128);
+
+/// A [`Plugin`] which creates a [`Checksum`] resource which can be read after or during the
+/// [`SaveWorldSet::Snapshot`] set in the [`SaveWorld`] schedule has been run.
+///
+/// To add you own data to this [`Checksum`], create an [`Entity`] with a [`ChecksumPart`]
+/// [`Component`]. Every [`Entity`] with this [`Component`] will participate in the
+/// creation of a [`Checksum`].
+
+pub struct GgrsChecksumPlugin;
+
+impl GgrsChecksumPlugin {
+    /// A [`System`] responsible for updating [`Checksum`] based on [`ChecksumParts`](`ChecksumPart`).
+    pub fn update(mut checksum: ResMut<Checksum>, parts: Query<&ChecksumPart>) {
+        // TODO: Add explicit ordering to `ChecksumPart`'s to make checksum more robust to transposition
```

Comment:
This is done now, right?

Reply:
If there are two things that contribute to the checksum (lets say a resource `Foo(usize)` and a resource `Bar(usize)`), their individual hashes will be identical for identical inner values (type information isn't passed into `Hash`). So if you have `Foo(5)` and `Bar(7)` on frame 5, then `Foo(7)` and `Bar(5)` on frame 6, frames 5 and 6 will have identical checksums, because `a ^ b == b ^ a`.

By having a guaranteed ordering, you could instead hash the whole sequence, which will produce unique checksums for frames 5 and 6 in the above example.

At first I thought about including `TypeId` or `type_name` in the hash, since we have access to that information and it's order independent, but those values aren't stable yet, so it's possible the checksum could just break for cross-platform multiplayer.

A workaround for that could be to have a `HashMap<TypeId, usize>` that's built as data types are added to the checksum process, but I don't know if the above transposition issue is even common enough to justify the added overhead.

Reply:
Ah, I see. Makes sense.

Reply:
We could create a constructor where we pass in a value and leave it up to calling code. Or it could be a const generic. Just ideas. Think I prefer the simpler API over more robust checksumming.